File: practical1/minstack.py
class MinStack:
    def __init__(self, capacity):
        self.capacity = capacity
        self.stack = []
        self.min_stack = []  # Auxiliary stack to track minimum elements
        
    def is_empty(self):
        return len(self.stack) == 0
        
    def is_full(self):
        return len(self.stack) == self.capacity
        
    def push(self, value):
        if self.is_full():
            print("Stack Overflow! Cannot push element.")
            return
            
        self.stack.append(value)
        
        # Update min_stack
        if not self.min_stack or value <= self.min_stack[-1]:
            self.min_stack.append(value)
        
        print(f"Pushed {value} to stack")
        
    def pop(self):
        if self.is_empty():
            print("Stack Underflow! Cannot pop element.")
            return None
            
        popped = self.stack.pop()
        
        # Update min_stack if we're popping the current minimum
        if popped == self.min_stack[-1]:
            self.min_stack.pop()
            
        print(f"Popped {popped} from stack")
        return popped
        
    def top(self):
        if self.is_empty():
            print("Stack is empty!")
            return None
        return self.stack[-1]
        
    def get_min(self):
        if self.is_empty():
            print("Stack is empty!")
            return None
        return self.min_stack[-1]
        
    def display(self):
        if self.is_empty():
            print("Stack is empty!")
            return
        print("Stack elements (top to bottom):", self.stack[::-1])
        print("Current minimum:", self.get_min())

# Example usage
if __name__ == "__main__":
    # Create a stack with capacity 5
    stack = MinStack(5)
    
    # Test operations
    print("Pushing elements...")
    stack.push(3)
    stack.push(5)
    stack.push(2)
    stack.push(1)
    
    print("\nDisplaying stack:")
    stack.display()
    
    print(f"\nTop element: {stack.top()}")
    print(f"Minimum element: {stack.get_min()}")
    
    print("\nPopping elements...")
    stack.pop()
    stack.pop()
    
    print("\nDisplaying stack after pops:")
    stack.display()
    
    # Test overflow
    print("\nTesting overflow...")
    stack.push(7)
    stack.push(8)
    stack.push(9)  # This should show overflow error
    
    # Test underflow
    print("\nTesting underflow...")
    stack.pop()
    stack.pop()
    stack.pop()
    stack.pop()
    stack.pop()  # This should show underflow error

File: practical1/minstack.js
class MinStack {
    constructor(capacity) {
        this.capacity = capacity;
        this.stack = [];
        this.minStack = [];
    }
    
    isEmpty() {
        return this.stack.length === 0;
    }
    
    isFull() {
        return this.stack.length === this.capacity;
    }
    
    push(value) {
        if (this.isFull()) {
            console.log("Stack Overflow! Cannot push element.");
            return;
        }
        
        this.stack.push(value);
        
        // Update minStack
        if (this.minStack.length === 0 || value <= this.minStack[this.minStack.length - 1]) {
            this.minStack.push(value);
        }
        
        console.log(`Pushed ${value} to stack`);
    }
    
    pop() {
        if (this.isEmpty()) {
            console.log("Stack Underflow! Cannot pop element.");
            return null;
        }
        
        const popped = this.stack.pop();
        
        // Update minStack if we're popping the current minimum
        if (popped === this.minStack[this.minStack.length - 1]) {
            this.minStack.pop();
        }
        
        console.log(`Popped ${popped} from stack`);
        return popped;
    }
    
    top() {
        if (this.isEmpty()) {
            console.log("Stack is empty!");
            return null;
        }
        return this.stack[this.stack.length - 1];
    }
    
    getMin() {
        if (this.isEmpty()) {
            console.log("Stack is empty!");
            return null;
        }
        return this.minStack[this.minStack.length - 1];
    }
    
    display() {
        if (this.isEmpty()) {
            console.log("Stack is empty!");
            return;
        }
        
        console.log("Stack elements (top to bottom):", [...this.stack].reverse());
        console.log("Current minimum:", this.getMin());
    }
}

// Example usage
function testMinStack() {
    // Create a stack with capacity 5
    const stack = new MinStack(5);
    
    // Test operations
    console.log("Pushing elements...");
    stack.push(3);
    stack.push(5);
    stack.push(2);
    stack.push(1);
    
    console.log("\nDisplaying stack:");
    stack.display();
    
    console.log(`\nTop element: ${stack.top()}`);
    console.log(`Minimum element: ${stack.getMin()}`);
    
    console.log("\nPopping elements...");
    stack.pop();
    stack.pop();
    
    console.log("\nDisplaying stack after pops:");
    stack.display();
    
    // Test overflow
    console.log("\nTesting overflow...");
    stack.push(7);
    stack.push(8);
    stack.push(9);  // This should show overflow error
    
    // Test underflow
    console.log("\nTesting underflow...");
    stack.pop();
    stack.pop();
    stack.pop();
    stack.pop();
    stack.pop();  // This should show underflow error
}

// Run the test
testMinStack();

